addons/web/static/src/views/fields/section_note_list/section_note_list_field.js

```diff
@@ -0,0 +1,387 @@
+                } else if (parent.type === DISPLAY_TYPES.SUBSECTION) {
+                    parent = parent.parent;
+                }
+            }
+            const item = new HierarchyItem(type, record, parent);
+            parent.addChild(item);
+            itemMap[item.id] = item;
+            if (item.isSection) {
+                parent = item;
+            }
+        }
+
+        return { hierarchy, itemMap };
+    }
+
+    /**
+     * @param {HierarchyItem} item
+     */
+    canAddSubSection(item) {
+        const displayTypeField = this.props.list.fields[this.props.displayTypeField];
+        const selection = new Map(displayTypeField.selection);
+        return selection.has(DISPLAY_TYPES.SUBSECTION) && item.isTopSection;
+    }
+
+    /**
+     * @param {HierarchyItem} item
+     */
+    async deleteSection(item) {
+        if (this.editedRecord && this.editedRecord !== item) {
+            const left = await this.props.list.leaveEditMode();
+            if (!left) {
+                return;
+            }
+        }
+        if (this.activeActions.onDelete) {
+            const method = this.props.isM2M ? "unlink" : "delete";
+            const commands = item.records.map((r) =>
+                x2ManyCommands[method](r.resId || r._virtualId)
+            );
+            await this.props.list.model.mutex.exec(async () => {
+                await this.props.list._applyCommands(commands);
+                await this.props.list._onUpdate();
+            });
+        }
+    }
+
+    /**
+     * @param {HierarchyItem} item
+     */
+    async duplicateSection(item) {
+        await this.props.list.model.mutex.exec(async () => {
+            const promises = [];
+            for (const record of item.records) {
+                const data = { ...record.data };
+                if (this.props.list.handleField) {
+                    data[this.props.list.handleField] = this.props.list.records.length + 1;
+                }
+                const copiedRecord = this.props.list._createRecordDatapoint(data, {
+                    virtualId: getId("virtual"),
+                    manuallyAdded: true,
+                });
+                const addingPromise = this.props.list._addRecord(copiedRecord);
+                promises.push(addingPromise);
+            }
+            await Promise.all(promises);
+            await this.props.list._onUpdate();
+        });
+    }
+
+    editNextRecord(record, group) {
+        const item = this.getHierarchyItem(record.id);
+        if (item.isSection || item.isLastChildItem) {
+            return this.addRowInSection(item, false);
+        } else {
+            return super.editNextRecord(record, group);
+        }
+    }
+
+    getHierarchyItem(id) {
+        return this.itemMap[id];
+    }
+
+    getListRecords(list) {
+        const { hierarchy, itemMap } = this.buildHierarchy(list);
+        this.itemMap = itemMap;
+        this.hierarchy = hierarchy;
+        return hierarchy.records;
+    }
```

Comment:
I'm not a fan of this, for several reasons (kind of a getter having side-effects, "imposing" this change in the ListRenderer, easy to get perf issue by computing stuff in a getter...).

As far as I understand, hierarchy.records is the list of records in the relation (which should already be properly ordered with sequences). So I would keep the `get records` as before, and call `buildHiearchy` in `onWillRender` => guaranteed to be called only once at each rendering